cpp/src/groupby/sort/group_merge_m2.cu

```diff
@@ -14,154 +14,90 @@
  * limitations under the License.
  */
 
-#include <cudf/column/column_device_view.cuh>
 #include <cudf/column/column_factories.hpp>
 #include <cudf/detail/aggregation/aggregation.hpp>
 #include <cudf/detail/valid_if.cuh>
-#include <cudf/dictionary/detail/iterator.cuh>
 #include <cudf/structs/structs_column_view.hpp>
-#include <cudf/utilities/memory_resource.hpp>
 #include <cudf/utilities/span.hpp>
 #include <cudf/utilities/type_dispatcher.hpp>
 
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
 
+#include <cuda/functional>
 #include <cuda/std/functional>
 #include <thrust/iterator/counting_iterator.h>
-#include <thrust/iterator/discard_iterator.h>
 #include <thrust/iterator/zip_iterator.h>
-#include <thrust/reduce.h>
 #include <thrust/transform.h>
 #include <thrust/tuple.h>
 
 namespace cudf {
 namespace groupby {
 namespace detail {
 namespace {
-/**
- * @brief Struct to store partial results for merging.
- */
-template <class result_type>
-struct partial_result {
-  size_type count;
-  result_type mean;
-  result_type M2;
-};
-
-/**
- * @brief Functor to accumulate (merge) all partial results corresponding to the same key into a
- * final result storing in a member variable. It performs merging for the partial results of
- * `COUNT_VALID`, `MEAN`, and `M2` at the same time.
- */
-template <class result_type>
-struct accumulate_fn {
-  partial_result<result_type> merge_vals;
 
-  void __device__ operator()(partial_result<result_type> const& partial_vals) noexcept
-  {
-    if (partial_vals.count == 0) { return; }
-
-    auto const n_ab  = merge_vals.count + partial_vals.count;
-    auto const delta = partial_vals.mean - merge_vals.mean;
-    merge_vals.M2 += partial_vals.M2 + (delta * delta) *
-                                         static_cast<result_type>(merge_vals.count) *
-                                         static_cast<result_type>(partial_vals.count) / n_ab;
-    merge_vals.mean =
-      (merge_vals.mean * merge_vals.count + partial_vals.mean * partial_vals.count) / n_ab;
-    merge_vals.count = n_ab;
-  }
-};
+using result_type = double;
+static_assert(
+  std::is_same_v<cudf::detail::target_type_t<result_type, aggregation::Kind::M2>, result_type>);
 
 /**
  * @brief Functor to merge partial results of `COUNT_VALID`, `MEAN`, and `M2` aggregations
  * for a given group (key) index.
  */
-template <class result_type>
+template <typename count_type>
 struct merge_fn {
-  size_type const* const d_offsets;
-  size_type const* const d_counts;
-  result_type const* const d_means;
-  result_type const* const d_M2s;
+  size_type const* d_offsets;
+  count_type const* d_counts;
+  result_type const* d_means;
+  result_type const* d_M2s;
 
-  auto __device__ operator()(size_type const group_idx) noexcept
+  auto __device__ operator()(size_type const group_idx) const
```

Comment:
👍 adding the `const`.

I would have thought `noexcept` is implied already from the `__device__`.